src/meta/Drag.ts

```diff
@@ -0,0 +1,170 @@
+import global from '@dojo/shim/global';
+import { assign } from '@dojo/shim/object';
+import WeakMap from '@dojo/shim/WeakMap';
+import { Base } from './Base';
+
+export interface DragResults {
+	/**
+	 * The movement of pointer during the duration of the drag state
+	 */
+	delta: Position;
+
+	/**
+	 * Is the DOM node currently in a drag state
+	 */
+	isDragging: boolean;
+}
+
+interface NodeData {
+	dragResults: DragResults;
+	invalidate: () => void;
+	last: Position;
+	start: Position;
+}
+
+export interface Position {
+	x: number;
+	y: number;
+}
+
+/**
+ * A frozen empty result object, frozen to ensure that no one downstream modifies it
+ */
+const emptyResults = Object.freeze({
+	delta: Object.freeze({ x: 0, y: 0 }),
+	isDragging: false
+});
+
+/**
+ * Return the x/y position for an event
+ * @param e The MouseEvent or TouchEvent
+ */
+function getPosition(e: PointerEvent): Position {
+	return {
+		x: e.pageX,
+		y: e.pageY
+	};
+}
+
+/**
+ * Return the delta position between two positions
+ * @param start The first posistion
+ * @param current The second posistion
+ */
+function getDelta(start: Position, current: Position): Position {
+	return {
+		x: current.x - start.x,
+		y: current.y - start.y
+	};
+}
+
+class DragController {
+	private _nodeMap = new WeakMap<HTMLElement, NodeData>();
+	private _dragging: HTMLElement | undefined = undefined;
+
+	private _getData(target: HTMLElement): { state: NodeData, target: HTMLElement } | undefined {
+		if (this._nodeMap.has(target)) {
+			return { state: this._nodeMap.get(target)!, target };
```

Comment:
it would be so good if TS could know that we are in a block narrowed by `this._nodeMap.has(target)`

Reply:
See: Microsoft/TypeScript#11117

Reply:
Not surprised there’s an issue. That would be nice.